distributed/worker.py

```diff
@@ -1555,10 +1586,104 @@ def transition(self, ts, finish, **kwargs):
             self.validate_task(ts)
         self._notify_plugins("transition", ts.key, start, state or finish, **kwargs)
 
-    def transition_waiting_flight(self, ts, worker=None):
+    def transition_new_waiting(self, ts):
         try:
             if self.validate:
-                assert ts.state != "flight"
+                assert ts.state == "new"
+                assert ts.runspec is not None
+                assert not ts.who_has
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_new_fetch(self, ts):
+        try:
+            if self.validate:
+                assert ts.state == "new"
+                assert ts.runspec is None
+
+            for dependent in ts.dependents:
+                dependent.waiting_for_data.add(ts.key)
+
+            self.data_needed.append(ts.key)
+            self.waiting_for_data_count += 1
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_fetch_waiting(self, ts, runspec=None):
+        """This is a rescheduling transition that occurs after a worker failure.
+        A task was available from another worker but that worker died and the
+        scheduler reassigned the task for computation here.
+        """
```

Comment:
Just a comment

I believe this is one of the crucial transitions which mess up the state occasionally. Especially since if this transition is done, chances are there is still a coroutine in the background trying to fetch the task. If, for instance, this transition was just triggered because a worker was unresponsive, that background task may still fetch the result and we're faced with a `fetch -> waiting -> memory` or worse 
```
fetch -> waiting     -> memory
                    -> executing -> memory
```

I believe there is little we can do about these things other than handling them properly. In particular this means we might have a result already in memory even though its task is executing still